### ISSUE TITLE
bigcode/tiny_starcoder_py · Hugging Face

### ISSUE THREAD

**Issue by irthomasthomas** [irthomasthomas/undecidability#324]
> **Note:**
> 
> [bigcode/tiny_starcoder_py · Hugging Face](https://huggingface.co/bigcode/tiny_starcoder_py)
> 
> TinyStarCoderPy
> 
> This is a 164M parameters model with the same architecture as StarCoder (8k context length, MQA & FIM). It was trained on the Python data from StarCoderData for ~6 epochs which amounts to 100B tokens.
> 
> Use
> 
> Intended use
> 
> The model was trained on GitHub code, to assist with some tasks like Assisted Generation. For pure code completion, we advise using our 15B models StarCoder or StarCoderBase.
> 
> Generation
> 
> ```python
> # pip install -q transformers
> from transformers import AutoModelForCausalLM, AutoTokenizer
> 
> checkpoint = "bigcode/tiny_starcoder_py"
> device = "cuda" # for GPU usage or "cpu" for CPU usage
> 
> tokenizer = AutoTokenizer.from_pretrained(checkpoint)
> model = AutoModelForCausalLM.from_pretrained(checkpoint).to(device)
> 
> inputs = tokenizer.encode("def print_hello_world():", return_tensors="pt").to(device)
> outputs = model.generate(inputs)
> print(tokenizer.decode(outputs[0]))
> ```
> 
> Fill-in-the-middle
> 
> Fill-in-the-middle uses special tokens to identify the prefix/middle/suffix part of the input and output:
> 
> ```python
> input_text = "<fim_prefix>def print_one_two_three():\n    print('one')\n    <fim_suffix>\n    print('three')<fim_middle>"
> inputs = tokenizer.encode(input_text, return_tensors="pt").to(device)
> outputs = model.generate(inputs)
> print(tokenizer.decode(outputs[0]))
> ```
> 
> Training
> 
> Model
> 
> - Architecture: GPT-2 model with multi-query attention and Fill-in-the-Middle objective
> - Pretraining steps: 50k
> - Pretraining tokens: 100 billion
> - Precision: bfloat16
> 
> Hardware
> 
> - GPUs: 32 Tesla A100
> - Training time: 18 hours
> 
> Software
> 
> - Orchestration: Megatron-LM
> - Neural networks: PyTorch
> - BP16 if applicable: apex
> 
> License
> 
> The model is licensed under the BigCode OpenRAIL-M v1 license agreement. You can find the full agreement [here](https://huggingface.co/bigcode/tiny_starcoder_py/blob/main/LICENSE).
> 
> #### Suggested labels
> 
> - { "key": "llm-pretraining", "value": "Information related to the pretraining process of Large Language Models" }

**Comment by irthomasthomas** [comment]
There is no technical reason why we can't use a second speculative decoder model. Making three models in total.

This model is 164M parameters. An order of magnitude larger would be 1.6B, and an order of magnitude again would still only be 16B parameters. At 8bit that makes about 18GB + the context window. I have a 12GB 3060, and an older, slower 8GB 1080. So this is doable in vram, but I'm not sure if that 1080 is actually faster than the GPU.
